sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollectCTEDefinitions.scala

```diff
@@ -282,11 +283,15 @@ object CTESubstitution extends Rule[LogicalPlan] {
           if (alwaysInline) {
             d.child
           } else {
-            // Add a `SubqueryAlias` for hint-resolving rules to match relation names.
-            SubqueryAlias(table, CTERelationRef(d.id, d.resolved, d.output, d.isStreaming))
+            // Add unresolved with CTE relations to the plan and we
+            // will do CTE resolution later in analyzer based on this node.
+            UnresolvedWithCTERelations(u, cteRelations)
```

Comment:
Thinking more about it, I think we can still return `SubqueryAlias(...)` as before here, which is like a shortcut as we know we should look up CTE relations first when resolving `UnresolvedRelation`. There is no need to delay it.

Reply:
So I should return here something like SubqueryAlias(table, UnresolvedRelation(...)) ?